README.rst
===============================
Markdown Twitter Extension
===============================

.. image:: https://badge.fury.io/py/mdx-twitter.png
    :target: http://badge.fury.io/py/mdx-twitter
    
.. image:: https://travis-ci.org/zen4ever/mdx-twitter.png?branch=master
        :target: https://travis-ci.org/zen4ever/mdx-twitter

.. image:: https://pypip.in/d/mdx-twitter/badge.png
        :target: https://crate.io/packages/mdx-twitter?version=latest


Markdown extension for embedding tweets using twitter OEmbed API

* Free software: BSD license

How to use
----------

Allows you to embed tweets into your Markdown.

Just install the package:

    .. code-block:: bash

        pip install mdx-twitter

Because Twitter API 1.1 requires you to authenticate, you will need to create
a config file with your credentials at '~/.mdx_twitter.cfg'::

    [Twitter]
    CONSUMER_KEY=xxxxxxxxxxxx
    CONSUMER_SECRET=xxxxxxxxxxxx
    ACCESS_TOKEN=xxxxxxxxxxxx
    ACCESS_TOKEN_SECRET=xxxxxxxxxxxx

Then just add 'twitter' to the list of your extensions:

    .. code-block:: python

        import markdown

        md = markdown.Markdown(extensions=['twitter'])

Twitter urls in your Makdown will become embedded tweets::

        https://twitter.com/jasoncosta/status/240192632003911681

Django integration
------------------

Since Twitter has a rate limit on their APIs, and doing network calls all the
time just to render Markdown is a bad idea anyways, we would need some caching.

If you are using `mdx-twitter` in a Django project, it will automatically
use your cache settings to cache HTML returned from Twitter API.

Also, you could place your API credentials in TWITTER_SETTINGS variable in your
project's `settings.py`, instead of using .cfg file

    .. code-block:: python

        TWITTER_SETTINGS = {
            'CONSUMER_KEY': '',
            'CONSUMER_SECRET': '',
            'ACCESS_TOKEN': '',
            'ACCESS_TOKEN_SECRET': '',
        }

Embedding tweets in UIWebView
-----------------------------

There are currently some problems with embedding tweets in UIWebView.

* Protocol agnostic src attribute "//platform.twitter.com/widgets.js" doesn't
  allow script to load

* There seems to be a problem with automatically determining width of the
  container

  https://dev.twitter.com/discussions/15450

So there is a 'width' configuration option that offers a quick 'fix' for those
problems. Just specify expected width of your tweet in pixels. 

    .. code-block:: python

        import markdown

        md = markdown.Markdown(extensions=['twitter(width=300)'])

It will only work with 'style' full.

Styles
------

You can customize appearance of your tweets using 'style' configuration
options. Default style is 'full', which includes script javascript, you could
also use 'simple' style, which just includes tweet in a blockquote tag without
javascript.

Another option would be to specify a path to your own function, which accepts
Twitter response json as an argument and should return final html.

An example custom style is included in the library. You can use it by
specifying 'style=mdx_twitter.custom_style'

TODO
----

* If we embed multiple tweets, we don't need to have `script` tag after each
  tweet. We should probably just have a `script` tag at the end of the
  document.


HISTORY.rst
.. :changelog:

History
-------

0.1.0 (2013-08-31)
++++++++++++++++++

* First release on PyPI.

tests/tests.py
import os
import re
import markdown
import json

from django.test import TestCase
from django.core.cache import cache

from mdx_twitter.extension import TWITTER_LINK_RE, get_cache_key

from httmock import urlmatch, HTTMock


@urlmatch(netloc=r'(.*\.)?twitter\.com$')
def twitter_mock(url, request):
    if 'omit_script' in request.url:
        return """{
            "html": "<blockquote class='twitter-tweet'><p>Need to plug in for a couple of hours? Here's the mix: 'NERO - Essential Mix (First broadcast Nov 2010)' - <a href='http://t.co/9MwHZCya' title='http://soundcloud.com/nerouk/nero-essential-mix-first?utm_source=soundcloud&amp;utm_campaign=share&amp;utm_medium=twitter&amp;utm_content=http://soundcloud.com/nerouk/nero-essential-mix-first'>soundcloud.com/nerouk/nero-es\\u2026</a></p>&mdash; Jason Costa (@jasoncosta) <a href='https://twitter.com/jasoncosta/status/240192632003911681' data-datetime='2012-08-27T21:02:40+00:00'>August 27, 2012</a></blockquote>\\n",
          "author_name": "Jason Costa",
          "provider_url": "http://twitter.com",
          "url": "https://twitter.com/jasoncosta/status/240192632003911681",
          "provider_name": "Twitter",
          "version": "1.0",
          "type": "rich",
          "height": null,
          "cache_age": "31536000000",
          "author_url": "https://twitter.com/jasoncosta",
          "width": 550}"""
    return """{
        "html": "<blockquote class='twitter-tweet'><p>Need to plug in for a couple of hours? Here's the mix: 'NERO - Essential Mix (First broadcast Nov 2010)' - <a href='http://t.co/9MwHZCya' title='http://soundcloud.com/nerouk/nero-essential-mix-first?utm_source=soundcloud&amp;utm_campaign=share&amp;utm_medium=twitter&amp;utm_content=http://soundcloud.com/nerouk/nero-essential-mix-first'>soundcloud.com/nerouk/nero-es\\u2026</a></p>&mdash; Jason Costa (@jasoncosta) <a href='https://twitter.com/jasoncosta/status/240192632003911681' data-datetime='2012-08-27T21:02:40+00:00'>August 27, 2012</a></blockquote>\\n<script src=\\"//platform.twitter.com/widgets.js\\" charset='utf-8'></script>",
      "author_name": "Jason Costa",
      "provider_url": "http://twitter.com",
      "url": "https://twitter.com/jasoncosta/status/240192632003911681",
      "provider_name": "Twitter",
      "version": "1.0",
      "type": "rich",
      "height": null,
      "cache_age": "31536000000",
      "author_url": "https://twitter.com/jasoncosta",
      "width": 550}"""


class TestDjangoInterations(TestCase):

    def test_django_caching(self):
        md = markdown.Markdown(extensions=['twitter'], safe_mode=True)
        key = get_cache_key({'id': '240192632003911681'})
        self.assertEqual(key, 'mdx_twitter:id:240192632003911681')
        with HTTMock(twitter_mock):
            html = md.convert('http://twitter.com/jasoncosta/statuses/240192632003911681')
            self.assertEqual(json.loads(cache.get(key))['html'], html)

    def tearDown(self):
        cache.clear()


class TestMdxTwitter(TestCase):

    def test_regex_1(self):
        m = re.match(TWITTER_LINK_RE, "https://twitter.com/jasoncosta/status/240192632003911681")
        self.assertEqual(m.group('tweet_id'), '240192632003911681')

    def test_regex_2(self):
        m = re.match(TWITTER_LINK_RE, "http://twitter.com/jasoncosta/statuses/240192632003911681")
        self.assertEqual(m.group('tweet_id'), '240192632003911681')

    def test_oembed(self):
        md = markdown.Markdown(extensions=['twitter'])

        with HTTMock(twitter_mock):
            html = md.convert('http://twitter.com/jasoncosta/statuses/240192632003911681')
            result = u"<blockquote class='twitter-tweet'><p>Need to plug in for a couple of hours? Here's the mix: 'NERO - Essential Mix (First broadcast Nov 2010)' - <a href='http://t.co/9MwHZCya' title='http://soundcloud.com/nerouk/nero-essential-mix-first?utm_source=soundcloud&amp;utm_campaign=share&amp;utm_medium=twitter&amp;utm_content=http://soundcloud.com/nerouk/nero-essential-mix-first'>soundcloud.com/nerouk/nero-es\u2026</a></p>&mdash; Jason Costa (@jasoncosta) <a href='https://twitter.com/jasoncosta/status/240192632003911681' data-datetime='2012-08-27T21:02:40+00:00'>August 27, 2012</a></blockquote>\n<script src=\"//platform.twitter.com/widgets.js\" charset='utf-8'></script>"  # NOQA
            self.assertEqual(html, result)

    def test_width(self):
        md = markdown.Markdown(extensions=['twitter(width=400)'])
        with HTTMock(twitter_mock):
            html = md.convert('http://twitter.com/jasoncosta/statuses/240192632003911681')
            self.assertNotEqual(html.find('400px'), -1)
            self.assertNotEqual(html.find('https://platform.twitter.com/'), -1)

    def test_style_simple(self):
        md = markdown.Markdown(extensions=['twitter(style=simple)'])
        with HTTMock(twitter_mock):
            html = md.convert('http://twitter.com/jasoncosta/statuses/240192632003911681')
            self.assertEqual(html.find('<script'), -1)

    def test_style_custom(self):
        md = markdown.Markdown(extensions=['twitter(style=mdx_twitter.custom_style)'])
        with HTTMock(twitter_mock):
            html = md.convert('http://twitter.com/jasoncosta/statuses/240192632003911681')
            self.assertEqual(html.find('<script'), -1)
            self.assertEqual(html, u'<p><a href="https://twitter.com/jasoncosta/status/240192632003911681">Jason Costa: </a>Need to plug in for a couple of hours? Here\'s the mix: \'NERO - Essential Mix (First broadcast Nov 2010)\' - <a href="http://t.co/9MwHZCya" title="http://soundcloud.com/nerouk/nero-essential-mix-first?utm_source=soundcloud&amp;utm_campaign=share&amp;utm_medium=twitter&amp;utm_content=http://soundcloud.com/nerouk/nero-essential-mix-first">soundcloud.com/nerouk/nero-es\u2026</a></p>')

    if os.path.exists(os.path.expanduser('~/.mdx_twitter.cfg')):
        def test_live_oembed(self):
            import markdown
            md = markdown.Markdown(extensions=['twitter'], safe_mode=True)
            html = md.convert('http://twitter.com/jasoncosta/statuses/240192632003911681')
            self.assertEqual(html, u'<blockquote class="twitter-tweet"><p>Need to plug in for a couple of hours? Here&#39;s the mix: &quot;NERO - Essential Mix (First broadcast Nov 2010)&quot; - <a href="http://t.co/9MwHZCya">http://t.co/9MwHZCya</a></p>&mdash; Jason Costa (@jasoncosta) <a href="https://twitter.com/jasoncosta/statuses/240192632003911681">August 27, 2012</a></blockquote>\n<script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script>')  # NOQA

    def tearDown(self):
        cache.clear()


tests/test_mdx_twitter.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_mdx-twitter
----------------------------------

Tests for `mdx-twitter` module.
"""

import unittest
import re
import os
from mdx_twitter.extension import TWITTER_LINK_RE

from httmock import urlmatch, HTTMock


@urlmatch(netloc=r'(.*\.)?twitter\.com$')
def twitter_mock(url, request):
    return """{
        "html": "<blockquote class='twitter-tweet'><p>Need to plug in for a couple of hours? Here's the mix: 'NERO - Essential Mix (First broadcast Nov 2010)' - <a href='http://t.co/9MwHZCya' title='http://soundcloud.com/nerouk/nero-essential-mix-first?utm_source=soundcloud&amp;utm_campaign=share&amp;utm_medium=twitter&amp;utm_content=http://soundcloud.com/nerouk/nero-essential-mix-first'>soundcloud.com/nerouk/nero-es\\u2026</a></p>&mdash; Jason Costa (@jasoncosta) <a href='https://twitter.com/jasoncosta/status/240192632003911681' data-datetime='2012-08-27T21:02:40+00:00'>August 27, 2012</a></blockquote>\\n<script src='//platform.twitter.com/widgets.js' charset='utf-8'></script>",
      "author_name": "Jason Costa",
      "provider_url": "http://twitter.com",
      "url": "https://twitter.com/jasoncosta/status/240192632003911681",
      "provider_name": "Twitter",
      "version": "1.0",
      "type": "rich",
      "height": null,
      "cache_age": "31536000000",
      "author_url": "https://twitter.com/jasoncosta",
      "width": 550}"""


class TestMdxTwitter(unittest.TestCase):

    def setUp(self):
        pass

    def test_regex_1(self):
        m = re.match(TWITTER_LINK_RE, "https://twitter.com/jasoncosta/status/240192632003911681")
        self.assertEqual(m.group('tweet_id'), '240192632003911681')

    def test_regex_2(self):
        m = re.match(TWITTER_LINK_RE, "http://twitter.com/jasoncosta/statuses/240192632003911681")
        self.assertEqual(m.group('tweet_id'), '240192632003911681')

    def test_oembed(self):
        import markdown
        md = markdown.Markdown(extensions=['twitter'])

        with HTTMock(twitter_mock):
            html = md.convert('http://twitter.com/jasoncosta/statuses/240192632003911681')
            self.assertEqual(html, u"<blockquote class='twitter-tweet'><p>Need to plug in for a couple of hours? Here's the mix: 'NERO - Essential Mix (First broadcast Nov 2010)' - <a href='http://t.co/9MwHZCya' title='http://soundcloud.com/nerouk/nero-essential-mix-first?utm_source=soundcloud&amp;utm_campaign=share&amp;utm_medium=twitter&amp;utm_content=http://soundcloud.com/nerouk/nero-essential-mix-first'>soundcloud.com/nerouk/nero-es\u2026</a></p>&mdash; Jason Costa (@jasoncosta) <a href='https://twitter.com/jasoncosta/status/240192632003911681' data-datetime='2012-08-27T21:02:40+00:00'>August 27, 2012</a></blockquote>\n<script src='//platform.twitter.com/widgets.js' charset='utf-8'></script>")  # NOQA

    @unittest.skipUnless(
        os.path.exists(os.path.expanduser('~/.mdx_twitter.cfg')),
        "~/.mdx_twitter.cfg is absent"
    )
    def test_live_oembed(self):
        import markdown
        md = markdown.Markdown(extensions=['twitter'], safe_mode=True)
        html = md.convert('http://twitter.com/jasoncosta/statuses/240192632003911681')
        self.assertEqual(html, u'<blockquote class="twitter-tweet"><p>Need to plug in for a couple of hours? Here&#39;s the mix: &quot;NERO - Essential Mix (First broadcast Nov 2010)&quot; - <a href="http://t.co/9MwHZCya">http://t.co/9MwHZCya</a></p>&mdash; Jason Costa (@jasoncosta) <a href="https://twitter.com/jasoncosta/statuses/240192632003911681">August 27, 2012</a></blockquote>\n<script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script>')  # NOQA

    def tearDown(self):
        pass

if __name__ == '__main__':
    unittest.main()


runtests.py
#!/usr/bin/env python

import sys
from os.path import abspath, dirname

from django.conf import settings

sys.path.insert(0, abspath(dirname(__file__)))

if not settings.configured:
    settings.configure(
        INSTALLED_APPS=(
            'django.contrib.contenttypes',
            'django.contrib.auth',
            'django.contrib.sessions',
            'tests',
        ),
        CACHES = {
            'default': {
                'BACKEND': 'django.core.cache.backends.locmem.LocMemCache',
            }
        },
        DATABASES={'default': {'ENGINE': 'django.db.backends.sqlite3'}},
    )


def runtests(modules=['tests']):
    from django.test.simple import DjangoTestSuiteRunner
    failures = DjangoTestSuiteRunner(failfast=False).run_tests(modules)
    sys.exit(failures)


if __name__ == '__main__':
    runtests()


mdx_twitter/__init__.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import logging

__author__ = 'Andrii Kurinnyi'
__email__ = 'andrew@marpasoft.com'
__version__ = '0.2.6'


def makeExtension(configs=None):
    from .extension import TwitterExtension
    from .settings import get_twitter_settings
    try:
        from django.conf import settings
        if settings.configured:
            from django.core.cache import cache
        else:
            cache = None
            logging.warning('Django settings are not configured')
    except ImportError:
        cache = None
    return TwitterExtension(
        configs=configs,
        twitter_settings=get_twitter_settings(),
        cache=cache
    )


def custom_style(result):
    from bs4 import BeautifulSoup
    soup = BeautifulSoup(result['html'])
    tweet = soup.find('blockquote').find('p')
    link = soup.new_tag("a")
    link['href'] = result['url']
    link.string = result['author_name'] + ': '
    tweet.insert(
        0, link
    )
    return unicode(tweet)


mdx_twitter/settings.py
import os
import logging
import ConfigParser


def get_twitter_settings():
    TWITTER_SETTINGS = {
        'CONSUMER_KEY': '',
        'CONSUMER_SECRET': '',
        'ACCESS_TOKEN': '',
        'ACCESS_TOKEN_SECRET': '',
    }

    # if we are in a Django environment,
    # let's try to get settings from there
    try:
        from django.conf import settings
        if settings.configured:
            twitter_settings = getattr(settings, 'TWITTER_SETTINGS', None)
            if twitter_settings:
                return twitter_settings
    except ImportError:
        logging.warning('Django is not installed')

    # we are not in Django, or TWITTER_SETTINGS variable is absent
    # let's try to get it from config file
    try:
        config = ConfigParser.ConfigParser()
        config.readfp(open(os.path.expanduser('~/.mdx_twitter.cfg')))
        for field in TWITTER_SETTINGS.keys():
            TWITTER_SETTINGS[field] = config.get('Twitter', field)
    except IOError:
        logging.warning('~/.mdx_twitter.cfg is missing')

    return TWITTER_SETTINGS


mdx_twitter/extension.py
import json
import logging

from markdown import Extension
from markdown.inlinepatterns import Pattern
from requests_oauthlib import OAuth1Session

TWITTER_LINK_RE = r"https?://twitter.com/([^/]+)/status(es)?/(?P<tweet_id>\d+)"


def get_cache_key(params):
    return ":".join(
        ['mdx_twitter'] + [item for sublist in params.items() for item in sublist]
    )


def process_tweet(result, tweet_style='full'):
    html = result['html']
    if tweet_style not in ['full', 'simple']:
        module, function = tweet_style.rsplit('.', 1)
        return getattr(__import__(module), function)(result)
    return html


class TwitterLinkPattern(Pattern):

    def get_html(self, tweet_id):
        params = {
            'id': tweet_id,
        }
        tweet_style = 'full'
        if self.ext.config['style'][0]:
            tweet_style = self.ext.config['style'][0]
        if tweet_style != 'full':
            params['omit_script'] = 'true'
        result = None
        if self.ext.cache:
            result = self.ext.cache.get(get_cache_key(params))
            if result:
                result = json.loads(result)
        if not result:
            response = self.ext.client.get(
                "https://api.twitter.com/1.1/statuses/oembed.json",
                params=params
            )
            result = response.json()
            if 'html' in result:
                if self.ext.cache:
                    self.ext.cache.set(
                        get_cache_key(params),
                        json.dumps(result),
                        timeout=60*60*24
                    )
            else:
                for err in result['errors']:
                    logging.warning(err['message'])
        if result and 'html' in result:
            html = process_tweet(result, tweet_style)
            # if we have width in a config,
            # we are probably use Markdow in a UIWebView
            # https://dev.twitter.com/discussions/15450
            if tweet_style == 'full' and self.ext.config['width'][0]:
                # let's fix the URL protocol
                html = html.replace(
                    '"//platform.twitter.com/',
                    '"https://platform.twitter.com/'
                )
                html = "<div style='width: %spx'>%s</div>" % (
                    self.ext.config['width'][0], html
                )
            return html
        return "[Tweet oembed error]"

    def handleMatch(self, m):
        tweet_id = m.group('tweet_id')
        html = self.get_html(tweet_id)
        placeholder = self.markdown.htmlStash.store(html, safe=True)
        return placeholder


class TwitterExtension(Extension):
    def __init__(self, configs, **kwargs):
        self.twitter_settings = kwargs.pop('twitter_settings')
        self.cache = kwargs.pop('cache')
        self.config = {
            'width': ['', 'Width of your tweet, useful for mobile'],
            'style': [
                '',
                'Style of the tweet: '
                'full, simple, or path to a function like "mdx_twitter.custom_style"'
            ],
        }
        for key, value in configs:
            self.setConfig(key, value)

    @property
    def client(self):
        if not hasattr(self, '_client'):
            self._client = OAuth1Session(
                self.twitter_settings['CONSUMER_KEY'],
                client_secret=self.twitter_settings['CONSUMER_SECRET'],
                resource_owner_key=self.twitter_settings['ACCESS_TOKEN'],
                resource_owner_secret=self.twitter_settings['ACCESS_TOKEN_SECRET']
            )
        return self._client

    def extendMarkdown(self, md, md_globals):
        pattern = TwitterLinkPattern(TWITTER_LINK_RE, md)
        pattern.ext = self
        md.inlinePatterns.add('twitter_link', pattern, '<reference')
